### DISCUSSION TITLE
Reconstructing full pseudorange from code phase for multiconstellations

### DISCUSSION THREAD

**Discussion by giaminhhoang** [JonasBchrt/snapshot-gnss-algorithms#4] [Q&A] [answered]
Hello,
In the snapshot positioning, Van Diggelen's algorithm is used to resolve ambiguous sub 1-ms pseudo-ranges for GPS L1 or 4-ms fo Galileo E1. In the case of both constellations, can we select 1 GPS satellite as a single reference and resolve for Galileo satellites by extending his algorithm as the 2 constellations are synchronized within 50 ns?
In the case of 5 GPS satellites and 1 Galileo satellite, does Galileo still help?
Thank you.

**Comment by JonasBchrt** [comment, marked as answer]
Hi Gia Minh,
I tried both options, (1) having a single reference satellite and a single common bias for all satellites and (2) having different ones for the individual satellite systems. Both options have their advantages and disadvantages. Option (2) should be more accurate and allows you to resolve initial position uncertainties greater than 100-150 km by considering Galileo's longer codes. On the other hand, it adds an additional time offset variable for each additional satellite system to the optimisation problem, requiring more satellites to solve it accurately. (So, 5 GPS + 1 Galileo shouldn't help much.) Option (1) should be less accurate because it neglects the time offsets between the systems and is only likely to convergence in the same convergence radius as GPS, i.e., 100-150 km. On the other hand, it requires less satellites and hence is more robust in scenarios with reduced sky visibility and/or weak received signals.
For our publication, I chose option (1) because it reflects better our strive for good robustness rather than good accuracy. The 50 ns offset that you mention corresponds to an error of the individual satellite-receiver distances of 15 m, most of which will be vertical. In contrast, our sampling frequency is just 4 MHz, corresponding to a satellite-receiver distance resolution of 75 m and hence can introduce errors of up to 37.5 m. That's why we neglect the former. But for other systems or goals, the trade-off might be a different one.
I hope that provides some insights ;)

**Comment by giaminhhoang** [comment]
Thank you for your very quick response! It all makes sense. Just a couple of questions more.

1) I understand that you are doing cloud-based navigation, which makes access to ephemeris easier. In the case of on-device, long-term ephemeris is one of the hard problems. If you have any good references for that, could you recommend them?
2) How could I buy a snapper board?

**Reply by JonasBchrt** [reply]
Cool.
Re 1: I am afraid that this is beyond my area of expertise. I know that several companies offer "extended" or "long-term" or "offline" ephemerides for this use case, which promise increased accuracy over up to about a month. For example, u-blox has that as part of their AssistNow service for up to 35 days. However, I don't have much experience with any of these services, at least not to the level that I could give recommendations.
Re 2: Thanks for your interest :) The SnapperGPS hardware is open source https://github.com/SnapperGPS/snappergps-pcb and licensed under a permissive license. So, in theory, you could download the production files, send it to a PCB manufacturer or assembly service, and get your own. However, this might be inconvenient for a single board. To address this and to the best of my knowledge, someone is currently organising a group purchasing campaign on https://groupgets.com/ , which would allow straightforward access to a single board (or a smaller quantity). The campaign is not online yet, but it should be later this month or next month. I can give you a heads-up here once I see that it is online.

**Reply by giaminhhoang** [reply]
Thanks a lot. Looking forward to getting a Snapper in hand!

**Reply by JonasBchrt** [reply]
OK, so, the group purchasing campaign will probably happen in a couple of month rather than now 😕 But we're distributing some SnapperGPS receivers to interested people these days, see [here](https://docs.google.com/forms/d/e/1FAIpQLScis5_MUQs3HNjfqwt6B9lzpPwzFNBm_w2HVoN9YH3vDohyuA/viewform?usp=sf_link). Your use case is perhaps not what we had in mind for this in the first place, but if we have enough stock, that would probably be fine. If you're still interested, you could have a look.

**Reply by giaminhhoang** [reply]
Thanks for your updates. I just filled the form. I believe I am a good tester for your case. Furthermore, I am developing also snappshot SW receiver and have experiences on navigation systems. My LinkedIn is https://www.linkedin.com/in/giaminhhoang/ including my resume, GitHub, and Google Scholar. Thanks for your consideration.
